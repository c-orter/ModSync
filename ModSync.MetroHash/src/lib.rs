#![no_main]

use js_sys::Uint8Array;
use metrohash::MetroHash128;
use std::hash::Hasher;
use std::mem;
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn metrohash128(data: &[u8]) -> Uint8Array {
    let mut hasher = MetroHash128::new();
    hasher.write(data);
    let hash: [u8; 16] = unsafe { mem::transmute(hasher.finish128()) };
    
    hash.as_slice().into()
}
