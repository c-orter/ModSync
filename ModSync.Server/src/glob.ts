/**
 * Shamelessly stolen from - https://github.com/aleclarson/glob-regex
 */

const dotRE = /\./g;
const dotPattern = "\\.";

const restRE = /\*\*$/g;
const restPattern = "(.+)";

// noinspection RegExpUnnecessaryNonCapturingGroup
const globRE = /(?:\*\*\/|\*\*|\*)/g;
const globPatterns: Record<string, string> = {
	"*": "([^/]+)", // no backslashes
	"**": "(.+/)?([^/]+)", // short for "**/*"
	"**/": "(.+/)?", // one or more directories
};

function mapToPattern(str: string) {
	return globPatterns[str];
}

function replace(glob: string) {
	return glob
		.replace(dotRE, dotPattern)
		.replace(restRE, restPattern)
		.replace(globRE, mapToPattern);
}

function join(globs: string[]) {
	return `((${globs.map(replace).join(")|(")}))`;
}

export function glob(glob: string | string[]) {
	return new RegExp(`^${Array.isArray(glob) ? join(glob) : replace(glob)}$`);
}

export function globNoEnd(glob: string | string[]) {
	return new RegExp(`^${Array.isArray(glob) ? join(glob) : replace(glob)}`);
}
