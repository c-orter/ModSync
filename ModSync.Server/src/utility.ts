import { cpus } from 'node:os';
import path from "node:path";


export class HttpError extends Error {
	constructor(
		public code: number,
		message: string,
	) {
		super(message);
	}

	get codeMessage(): string {
		switch (this.code) {
			case 400:
				return "Bad Request";
			case 404:
				return "Not Found";
			default:
				return "Internal Server Error";
		}
	}
}

export function winPath(p: string): string {
	return p.split(path.posix.sep).join(path.win32.sep);
}

export function unixPath(p: string): string {
	return p.split(path.win32.sep).join(path.posix.sep);
}

/**
 * A lock that is granted when calling [[Semaphore.acquire]].
 */
type Lock = {
	release: () => void
}

/**
 * A task that has been scheduled with a [[Semaphore]] but not yet started.
 */
type WaitingPromise = {
	resolve: (lock: Lock) => void
	reject: (err?: Error) => void
}

/**
 * A [[Semaphore]] is a tool that is used to control concurrent access to a common resource. This implementation
 * is used to apply a max-parallelism threshold.
 */
export class Semaphore {
	private running = 0
	private waiting: WaitingPromise[] = []

	constructor(public max: number = cpus().length) {
		if (max < 1) {
			throw new Error(
				`Semaphore was created with a max value of ${max} but the max value cannot be less than 1`,
			)
		}
	}

	private take() {
		if (this.waiting.length > 0 && this.running < this.max) {
			this.running++

			// Get the next task from the queue
			const task = this.waiting.shift()!

			// Resolve the promise to allow it to start, provide a release function
			task.resolve({ release: this.release })
		}
	}

	public acquire(): Promise<Lock> {
		if (this.running < this.max) {
			this.running++
			return Promise.resolve({ release: this.release })
		}

		return new Promise<Lock>((resolve, reject) => {
			this.waiting.push({ resolve, reject })
		})
	}

	private release = () => {
		this.running--
		this.take()
	}

	/**
	 * Purge all waiting tasks from the [[Semaphore]]
	 */
	public purge() {
		this.waiting.forEach(task => {
			task.reject(
				new Error('The semaphore was purged and as a result this task has been cancelled'),
			)
		})

		this.running = 0
		this.waiting = []
	}
}