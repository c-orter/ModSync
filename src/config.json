{
    "clientDirs": ["BepInEx/plugins", "BepInEx/config"],
    "serverDirs": ["user/mods"]
}