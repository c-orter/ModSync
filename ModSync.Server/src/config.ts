import path from "node:path";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { JsonUtil } from "@spt/utils/JsonUtil";
import type { VFS } from "@spt/utils/VFS";
import { glob, globNoEnd } from "./glob";
import { unixPath } from "./utility";
import type { ILogger } from "@spt/models/spt/utils/ILogger";

export type SyncPath = {
	path: string;
	enabled?: boolean;
	enforced?: boolean;
	silent?: boolean;
	restartRequired?: boolean;
};

type RawConfig = {
	syncPaths: (string | SyncPath)[];
	commonModExclusions: string[];
};

export class Config {
	constructor(
		public syncPaths: Required<SyncPath>[],
		public commonModExclusions: string[],
	) {}

	public isExcluded(filePath: string): boolean {
		return this.commonModExclusions.some((exclusion) =>
			glob(exclusion).test(unixPath(filePath)),
		);
	}

	public isParentExcluded(filePath: string): boolean {
		return this.commonModExclusions.some((exclusion) =>
			globNoEnd(exclusion).test(unixPath(filePath)),
		);
	}
}
export class ConfigUtil {
	constructor(
		private vfs: VFS,
		private jsonUtil: JsonUtil,
		private modImporter: PreSptModLoader,
		private logger: ILogger,
	) {}

	/**
	 * @throws {Error} If the config file does not exist
	 */
	private readConfigFile(): RawConfig {
		const modPath = this.modImporter.getModPath("Corter-ModSync");
		const configPath = path.join(modPath, "src", "config.jsonc");

		return this.jsonUtil.deserializeJsonC(
			this.vfs.readFile(configPath),
			"config.jsonc",
		) as RawConfig;
	}

	/**
	 * @throws {Error} If the config is invalid
	 */
	private validateConfig(config: RawConfig): void {
		if (!Array.isArray(config.syncPaths))
			throw new Error(
				"Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.",
			);

		if (!Array.isArray(config.commonModExclusions))
			throw new Error(
				"Corter-ModSync: config.jsonc 'commonModExclusions' is not an array. Please verify your config is correct and try again.",
			);

		for (const syncPath of config.syncPaths) {
			if (typeof syncPath !== "string" && !("path" in syncPath))
				throw new Error(
					"Corter-ModSync: config.jsonc 'syncPaths' is missing 'path'. Please verify your config is correct and try again.",
				);

			if (
				typeof syncPath === "string"
					? path.isAbsolute(syncPath)
					: path.isAbsolute(syncPath.path)
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must be relative to SPT server root. Invalid path '${syncPath}'`,
				);

			if (
				path
					.relative(
						process.cwd(),
						path.resolve(
							process.cwd(),
							typeof syncPath === "string" ? syncPath : syncPath.path,
						),
					)
					.startsWith("..")
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must within SPT server root. Invalid path '${syncPath}'`,
				);
		}
	}

	public load(): Config {
		const rawConfig = this.readConfigFile();
		this.validateConfig(rawConfig);

		return new Config(
			rawConfig.syncPaths.map((syncPath) => ({
				enabled: true,
				// Not yet implemented
				enforced: false,
				silent: false,
				restartRequired: true,
				...(typeof syncPath === "string" ? { path: syncPath } : syncPath),
			})),
			rawConfig.commonModExclusions,
		);
	}
}
