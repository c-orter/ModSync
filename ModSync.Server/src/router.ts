import type { HttpFileUtil } from "@spt/utils/HttpFileUtil";
import type { SyncUtil } from "./sync";
import { glob } from "./glob";
import type { IncomingMessage, ServerResponse } from "node:http";
import path from "node:path";
import type { VFS } from "@spt/utils/VFS";
import type { Config } from "./config";
import { HttpError, winPath } from "./utility";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { HttpServerHelper } from "@spt/helpers/HttpServerHelper";

const FALLBACK_SYNCPATHS: Record<string, object> = {
	undefined: ["BepInEx\\plugins\\Corter-ModSync.dll", "ModSync.Updater.exe"],
};

const FALLBACK_HASHES: Record<string, object> = {
	undefined: {
		"BepInEx\\plugins\\Corter-ModSync.dll": { crc: 999999999 },
		"ModSync.Updater.exe": { crc: 999999999 },
	},
};

export class Router {
	constructor(
		private config: Config,
		private syncUtil: SyncUtil,
		private vfs: VFS,
		private httpFileUtil: HttpFileUtil,
		private httpServerHelper: HttpServerHelper,
		private modImporter: PreSptModLoader,
		private logger: ILogger,
	) {}

	/**
	 * @internal
	 */
	public async getServerVersion(
		req: IncomingMessage,
		res: ServerResponse,
		_: RegExpMatchArray,
	) {
		const modPath = this.modImporter.getModPath("Corter-ModSync");
		const packageJson = JSON.parse(
			// @ts-expect-error readFile returns a string when given a valid encoding
			await this.vfs
				// @ts-expect-error readFile takes in an options object, including an encoding option
				.readFilePromisify(path.join(modPath, "package.json"), {
					encoding: "utf-8",
				}),
		);

		res.setHeader("Content-Type", "application/json");
		res.writeHead(200, "OK");
		res.end(JSON.stringify(packageJson.version));
	}

	/**
	 * @internal
	 */
	public async getSyncPaths(
		req: IncomingMessage,
		res: ServerResponse,
		_: RegExpMatchArray,
	) {
		const version = req.headers["modsync-version"] as string;
		if (version in FALLBACK_SYNCPATHS) {
			res.setHeader("Content-Type", "application/json");
			res.writeHead(200, "OK");
			res.end(JSON.stringify(FALLBACK_SYNCPATHS[version]));
			return;
		}

		res.setHeader("Content-Type", "application/json");
		res.writeHead(200, "OK");
		res.end(
			JSON.stringify(
				this.config.syncPaths.map(({ path, ...rest }) => ({
					path: winPath(path),
					...rest,
				})),
			),
		);
	}

	/**
	 * @internal
	 */
	public async getHashes(
		req: IncomingMessage,
		res: ServerResponse,
		_: RegExpMatchArray,
	) {
		const version = req.headers["modsync-version"] as string;
		if (version in FALLBACK_HASHES) {
			res.setHeader("Content-Type", "application/json");
			res.writeHead(200, "OK");
			res.end(JSON.stringify(FALLBACK_HASHES[version]));
			return;
		}

		console.time("hash");
		res.setHeader("Content-Type", "application/json");
		res.writeHead(200, "OK");
		res.end(
			JSON.stringify(await this.syncUtil.hashModFiles(this.config.syncPaths)),
		);
		console.timeEnd("hash");
	}

	/**
	 * @internal
	 */
	public async fetchModFile(
		_: IncomingMessage,
		res: ServerResponse,
		matches: RegExpMatchArray,
	) {
		const filePath = decodeURIComponent(matches[1]);

		const sanitizedPath = this.syncUtil.sanitizeDownloadPath(
			filePath,
			this.config.syncPaths,
		);

		if (!this.vfs.exists(sanitizedPath))
			throw new HttpError(
				404,
				`Attempt to access non-existent path ${filePath}`,
			);

		try {
			const fileStats = await this.vfs.statPromisify(sanitizedPath);
			res.setHeader(
				"Content-Type",
				this.httpServerHelper.getMimeText(path.extname(filePath)) ||
					"text/plain",
			);
			res.setHeader("Content-Length", fileStats.size);
			this.httpFileUtil.sendFile(res, sanitizedPath);
		} catch (e) {
			throw new HttpError(
				500,
				`Corter-ModSync: Error reading '${filePath}'\n${e}`,
			);
		}
	}

	public handleRequest(req: IncomingMessage, res: ServerResponse) {
		const routeTable = [
			{
				route: glob("/modsync/version"),
				handler: this.getServerVersion.bind(this),
			},
			{
				route: glob("/modsync/paths"),
				handler: this.getSyncPaths.bind(this),
			},
			{
				route: glob("/modsync/hashes"),
				handler: this.getHashes.bind(this),
			},
			{
				route: glob("/modsync/fetch/**"),
				handler: this.fetchModFile.bind(this),
			},
		];

		try {
			for (const { route, handler } of routeTable) {
				const matches = route.exec(req.url || "");
				if (matches) return handler(req, res, matches);
			}

			throw new HttpError(404, "Corter-ModSync: Unknown route");
		} catch (e) {
			if (e instanceof Error)
				this.logger.error(
					`Corter-ModSync: Error when handling [${req.method} ${req.url}]:\n${e.message}\n${e.stack}`,
				);

			if (e instanceof HttpError) {
				res.writeHead(e.code, e.codeMessage);
				res.end(e.message);
			} else {
				res.writeHead(500, "Internal server error");
				res.end(
					`Corter-ModSync: Error handling [${req.method} ${req.url}]:\n${e}`,
				);
			}
		}
	}
}
