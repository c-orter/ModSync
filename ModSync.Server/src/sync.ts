import type { VFS } from "@spt/utils/VFS";
import path from "node:path";
import { crc32Init, crc32Update, crc32Final } from "./crc";
import type { Config, SyncPath } from "./config";
import { HttpError, winPath } from "./utility";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import { createReadStream } from "node:fs";

type ModFile = {
	crc: number;
	nosync: boolean;
};

export class SyncUtil {
	constructor(
		private vfs: VFS,
		private config: Config,
		private logger: ILogger,
	) {}

	private async getFilesInDir(dir: string): Promise<[string, boolean][]> {
		if (!this.vfs.exists(dir)) {
			this.logger.warning(
				`Corter-ModSync: Directory '${dir}' does not exist, will be ignored.`,
			);
			return [];
		}

		const stats = await this.vfs.statPromisify(dir);
		if (stats.isFile())
			return [
				[
					dir,
					this.config.isExcluded(dir) ||
						this.vfs.exists(path.join(dir, ".nosync")) ||
						this.vfs.exists(path.join(dir, ".nosync.txt")),
				],
			];

		const nosyncDir =
			this.config.isExcluded(dir) ||
			this.vfs.exists(path.join(dir, ".nosync")) ||
			this.vfs.exists(path.join(dir, ".nosync.txt"));

		return (
			await Promise.all(
				this.vfs
					.getFiles(dir)
					.filter(
						(file) =>
							!file.endsWith(".nosync") && !file.endsWith(".nosync.txt"),
					)
					.map(
						async (file): Promise<[string, boolean]> => [
							path.join(dir, file),
							nosyncDir ||
								this.config.isExcluded(path.join(dir, file)) ||
								this.vfs.exists(`${path.join(dir, file)}.nosync`) ||
								this.vfs.exists(`${path.join(dir, file)}.nosync.txt`),
						],
					),
			)
		).concat(
			(
				await Promise.all(
					this.vfs
						.getDirs(dir)
						.map((subDir) => this.getFilesInDir(path.join(dir, subDir))),
				)
			)
				.flat()
				.map(([child, nosync]): [string, boolean] => [
					child,
					nosyncDir || nosync,
				]),
		);
	}

	private async buildModFile(
		file: string,
		// biome-ignore lint/correctness/noEmptyPattern: <explanation>
		{}: Required<SyncPath>,
		nosync: boolean,
	): Promise<ModFile> {
		try {
			return {
				nosync,
				crc: nosync ? 0 : await new Promise<number>((resolve, reject) => {
					let crc = crc32Init();

					createReadStream(file)
						.on("error", reject)
						.on("data", (data: Buffer) => {
							crc = crc32Update(crc, data);
						})
						.on("end", () => {
							resolve(crc32Final(crc));
						});
				}),
			};
		} catch (e) {
			throw new HttpError(500, `Corter-ModSync: Error reading '${file}'\n${e}`);
		}
	}

	public async hashModFiles(
		syncPaths: Config["syncPaths"],
	): Promise<Record<string, Record<string, ModFile>>> {
		return Object.fromEntries(
			await Promise.all(
				syncPaths.map(async (syncPath) => [
					winPath(syncPath.path),
					Object.fromEntries(
						await Promise.all(
							(await this.getFilesInDir(syncPath.path)).map(
								async ([file, nosync]) =>
									[
										winPath(file),
										await this.buildModFile(file, syncPath, nosync),
									] as const,
							),
						),
					),
				]),
			),
		);
	}

	/**
	 * @throws {Error} If file path is invalid
	 */
	public sanitizeDownloadPath(
		file: string,
		syncPaths: Config["syncPaths"],
	): string {
		const normalized = path.join(
			path.normalize(file).replace(/^(\.\.(\/|\\|$))+/, ""),
		);

		if (
			!syncPaths.some(
				({ path: p }) =>
					!path
						.relative(path.join(process.cwd(), p), normalized)
						.startsWith(".."),
			)
		)
			throw new HttpError(
				400,
				`Corter-ModSync: Requested file '${file}' is not in an enabled sync path!`,
			);

		return normalized;
	}
}
