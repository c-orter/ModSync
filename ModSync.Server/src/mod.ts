import type { DependencyContainer } from "tsyringe";

import type { IncomingMessage, ServerResponse } from "node:http";
import type { IPreSptLoadMod } from "@spt/models/external/IPreSptLoadMod";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import type { HttpListenerModService } from "@spt/services/mod/httpListener/HttpListenerModService";
import type { HttpFileUtil } from "@spt/utils/HttpFileUtil";
import type { FileSystem } from "@spt/utils/FileSystem";
import type { JsonUtil } from "@spt/utils/JsonUtil";
import { ConfigUtil, type Config } from "./config";
import { SyncUtil } from "./sync";
import { Router } from "./router";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { HttpServerHelper } from "@spt/helpers/HttpServerHelper";
import { Statter } from "./utility/statter";

class Mod implements IPreSptLoadMod {
	private static container: DependencyContainer;

	private static loadFailed = false;
	private static config: Config;

	public async preSptLoad(container: DependencyContainer): Promise<void> {
		Mod.container = container;
		const logger = container.resolve<ILogger>("WinstonLogger");
		const vfs = container.resolve<FileSystem>("FileSystem");
		const jsonUtil = container.resolve<JsonUtil>("JsonUtil");
		const modImporter = container.resolve<PreSptModLoader>("PreSptModLoader");
		const configUtil = new ConfigUtil(vfs, jsonUtil, modImporter, logger);
		const httpListenerService = container.resolve<HttpListenerModService>(
			"HttpListenerModService",
		);

		httpListenerService.registerHttpListener(
			"ModSyncListener",
			this.canHandleOverride,
			this.handleOverride,
		);

		try {
			Mod.config = await configUtil.load();
		} catch (e) {
			Mod.loadFailed = true;
			logger.error("Corter-ModSync: Failed to load config!");
			throw e;
		}

		if (!vfs.exists("ModSync.Updater.exe")) {
			Mod.loadFailed = true;
			logger.error(
				"Corter-ModSync: ModSync.Updater.exe not found! Please ensure ALL files from the release zip are extracted onto the server.",
			);
		}

		if (!vfs.exists("BepInEx/plugins/Corter-ModSync.dll")) {
			Mod.loadFailed = true;
			logger.error(
				"Corter-ModSync: Corter-ModSync.dll not found! Please ensure ALL files from the release zip are extracted onto the server.",
			);
		}
	}

	public canHandleOverride(_sessionId: string, req: IncomingMessage): boolean {
		return !Mod.loadFailed && (req.url?.startsWith("/modsync/") ?? false);
	}

	public async handleOverride(
		_sessionId: string,
		req: IncomingMessage,
		res: ServerResponse,
	): Promise<void> {
		const logger = Mod.container.resolve<ILogger>("WinstonLogger");
		const vfs = Mod.container.resolve<FileSystem>("FileSystem");
		const httpFileUtil = Mod.container.resolve<HttpFileUtil>("HttpFileUtil");
		const httpServerHelper =
			Mod.container.resolve<HttpServerHelper>("HttpServerHelper");
		const modImporter =
			Mod.container.resolve<PreSptModLoader>("PreSptModLoader");
		const statter = new Statter();
		const syncUtil = new SyncUtil(vfs, statter, Mod.config, logger);
		const router = new Router(
			Mod.config,
			syncUtil,
			vfs,
			statter,
			httpFileUtil,
			httpServerHelper,
			modImporter,
			logger,
		);

		try {
			router.handleRequest(req, res);
		} catch (e) {
			logger.error("Corter-ModSync: Failed to handle request!");
			throw e;
		}
	}
}

export const mod = new Mod();
