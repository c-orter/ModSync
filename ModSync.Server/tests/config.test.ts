import { expect, beforeEach, describe, it, vi } from "vitest";

import { Config, ConfigUtil } from "../src/config";
import { VFS } from "./utils/vfs";
import { JsonUtil } from "./utils/jsonUtil";
import { PreSptModLoader } from "./utils/preSptModLoader";
import type { VFS as IVFS } from "@spt/utils/VFS";
import type { JsonUtil as IJsonUtil } from "@spt/utils/JsonUtil";
import type { PreSptModLoader as IPreSptModLoader } from "@spt/loaders/PreSptModLoader";
import { fs, vol } from "memfs";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import { mock } from "vitest-mock-extended";
import path from "node:path";

vi.mock("node:fs", async () => (await vi.importActual("memfs")).fs);

describe("Config", () => {
	let config: Config;
	beforeEach(() => {
		config = new Config(
			[
				{
					path: "plugins",
					enabled: true,
					enforced: false,
					silent: false,
					restartRequired: true,
				},
				{
					path: "mods",
					enabled: false,
					enforced: false,
					silent: false,
					restartRequired: false,
				},
			],
			["plugins/**/node_modules", "plugins/**/*.js"],
		);
	});

	it("should correctly identify excluded paths", () => {
	
		expect(config.isExcluded("plugins/test.dll")).toBe(false);
		expect(config.isExcluded("plugins/banana/node_modules")).toBe(true);
		expect(config.isExcluded("plugins/banana/test.js")).toBe(true);
		expect(config.isExcluded("plugins/banana/config.json")).toBe(false);
		expect(config.isExcluded("plugins/banana/node_modules/lodash")).toBe(false);
	});
});

describe("ConfigUtil", () => {
	beforeEach(() => {
		vol.reset();
	});

	it("should create config if none exists", async () => {
		vol.fromNestedJSON(
			{ "placeholder.txt": "to ensure directory exists" },
			process.cwd(),
		);

		const config = await new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		).load();

		expect(fs.existsSync(path.join(process.cwd(), "config.jsonc"))).toBe(true);
	});

	it("should load config", async () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
				"syncPaths": [
					"plugins",
					{ "path": "mods", "enabled": false },
					{ "path": "doesnotexist", "enabled": true }
				],
				// Exclusions for commonly used SPT mods
				"exclusions": [
					"plugins/**/node_modules"
				]
			}`,
				plugins: {},
				mods: {},
			},
			process.cwd(),
		);

		const config = await new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		).load();

		expect(config.syncPaths).toEqual([
			{
				enabled: true,
				enforced: false,
				path: "doesnotexist",
				restartRequired: true,
				silent: false,
			},
			{
				enabled: true,
				enforced: false,
				path: "plugins",
				restartRequired: true,
				silent: false,
			},
			{
				enabled: false,
				enforced: false,
				path: "mods",
				restartRequired: true,
				silent: false,
			},
		]);

		expect(config.exclusions).toEqual(["plugins/**/node_modules"]);
	});

	it("should reject syncPath object without path", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						{ "enabled": true, "enforced": false, "path": "plugins", "restartRequired": true, "silent": false },
						{ "enabled": false, "enforced": false, "path": "mods", "restartRequired": true, "silent": false },
						{ "enabled": true, "enforced": false, "restartRequired": true, "silent": false }
					],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject absolute paths", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						"/etc/shadow",
						"C:\\\\Windows\\\\System32\\\\cmd.exe"
					],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject paths outside of SPT root", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						"../../etc/shadow"
					],
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject non-array syncPaths", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": "plugins",
					// Exclusions for commonly used SPT mods
					"exclusions": [
						"plugins/**/node_modules"
					]
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject on non-array exclusions", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"syncPaths": [
						"plugins"
					],
					// Exclusions for commonly used SPT mods
					"exclusions": "plugins/**/node_modules"
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});

	it("should reject invalid JSON", () => {
		vol.fromNestedJSON(
			{
				"config.jsonc": `{
					"invalid": "invalid"
				}`,
			},
			process.cwd(),
		);

		const configUtil = new ConfigUtil(
			new VFS() as IVFS,
			new JsonUtil() as IJsonUtil,
			new PreSptModLoader() as IPreSptModLoader,
			mock<ILogger>(),
		);

		expect(configUtil.load()).rejects.toThrowErrorMatchingSnapshot();
	});
});
