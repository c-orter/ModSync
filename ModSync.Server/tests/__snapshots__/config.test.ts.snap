// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfigUtil > Should always include ModSync.Updater.exe in syncPaths 1`] = `
[
  {
    "enabled": true,
    "enforced": true,
    "name": "[Required] ModSync Updater",
    "path": "ModSync.Updater.exe",
    "restartRequired": false,
    "silent": true,
  },
  {
    "enabled": true,
    "enforced": true,
    "name": "[Required] ModSync Plugin",
    "path": "BepInEx/plugins/Corter-ModSync.dll",
    "restartRequired": true,
    "silent": true,
  },
]
`;

exports[`ConfigUtil > should load config 1`] = `
[
  {
    "enabled": true,
    "enforced": true,
    "name": "[Required] ModSync Updater",
    "path": "ModSync.Updater.exe",
    "restartRequired": false,
    "silent": true,
  },
  {
    "enabled": true,
    "enforced": true,
    "name": "[Required] ModSync Plugin",
    "path": "BepInEx/plugins/Corter-ModSync.dll",
    "restartRequired": true,
    "silent": true,
  },
  {
    "enabled": true,
    "enforced": false,
    "name": "doesnotexist",
    "path": "doesnotexist",
    "restartRequired": true,
    "silent": false,
  },
  {
    "enabled": true,
    "enforced": false,
    "name": "plugins",
    "path": "plugins",
    "restartRequired": true,
    "silent": false,
  },
  {
    "enabled": false,
    "enforced": false,
    "name": "mods",
    "path": "mods",
    "restartRequired": true,
    "silent": false,
  },
]
`;

exports[`ConfigUtil > should reject absolute paths 1`] = `[Error: Corter-ModSync: SyncPaths must be relative to SPT server root. Invalid path '/etc/shadow']`;

exports[`ConfigUtil > should reject invalid JSON 1`] = `[Error: Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should reject non-array syncPaths 1`] = `[Error: Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should reject on non-array exclusions 1`] = `[Error: Corter-ModSync: config.jsonc 'exclusions' is not an array. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should reject on non-string syncPath name 1`] = `[Error: Corter-ModSync: config.jsonc 'syncPaths.name' must be a string. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should reject on syncPath name with invalid characters 1`] = `[JSONError: Bad control character in string literal in JSON at position 69 (line 3 column 48) while parsing '{					"syncPaths": [						{ "path": "p']`;

exports[`ConfigUtil > should reject paths outside of SPT root 1`] = `[Error: Corter-ModSync: SyncPaths must within SPT server root. Invalid path '../../etc/shadow']`;

exports[`ConfigUtil > should reject syncPath object without path 1`] = `[Error: Corter-ModSync: config.jsonc 'syncPaths' is missing 'path'. Please verify your config is correct and try again.]`;

exports[`ConfigUtil > should warn on explicitly excluded syncPaths 1`] = `[Error: Corter-ModSync: 'plugins' has been added as a sync path and is also in the 'exclusions' array. This probably isn't doing what you want. If you no longer want to sync this path, remove it from the 'exclusions' and 'syncPaths' arrays.]`;
